File: leetcode/AIcoding/2.Gemini-2.0/c_solutions/576-out-of-boundary-paths.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int findPaths(int m, int n, int maxMove, int startRow, int startColumn) {
    int dp[maxMove + 1][m][n];
    for (int k = 0; k <= maxMove; k++) {
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                dp[k][i][j] = 0;
            }
        }
    }
    int mod = 1000000007;
    int directions[4][2] = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
    for (int k = 1; k <= maxMove; k++) {
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                for (int d = 0; d < 4; d++) {
                    int prev_i = i - directions[d][0];
                    int prev_j = j - directions[d][1];
                    if (prev_i < 0 || prev_i >= m || prev_j < 0 || prev_j >= n) {
                        dp[k][i][j] = (dp[k][i][j] + 1) % mod;
                    } else {
                        dp[k][i][j] = (dp[k][i][j] + dp[k - 1][prev_i][prev_j]) % mod;
                    }
                }
            }
        }
    }
    return dp[maxMove][startRow][startColumn];
}

File: leetcode/AIcoding/2.Gemini-2.0/go_solutions/576-out-of-boundary-paths.go
func findPaths(m int, n int, maxMove int, startRow int, startColumn int) int {
    dp := make([][][]int, maxMove+1)
    for i := 0; i <= maxMove; i++ {
        dp[i] = make([][]int, m)
        for j := 0; j < m; j++ {
            dp[i][j] = make([]int, n)
        }
    }

    mod := 1000000007

    for i := 1; i <= maxMove; i++ {
        for r := 0; r < m; r++ {
            for c := 0; c < n; c++ {
                if r == 0 {
                    dp[i][r][c] = (dp[i][r][c] + 1) % mod
                }
                if r == m-1 {
                    dp[i][r][c] = (dp[i][r][c] + 1) % mod
                }
                if c == 0 {
                    dp[i][r][c] = (dp[i][r][c] + 1) % mod
                }
                if c == n-1 {
                    dp[i][r][c] = (dp[i][r][c] + 1) % mod
                }

                if r > 0 {
                    dp[i][r][c] = (dp[i][r][c] + dp[i-1][r-1][c]) % mod
                }
                if r < m-1 {
                    dp[i][r][c] = (dp[i][r][c] + dp[i-1][r+1][c]) % mod
                }
                if c > 0 {
                    dp[i][r][c] = (dp[i][r][c] + dp[i-1][r][c-1]) % mod
                }
                if c < n-1 {
                    dp[i][r][c] = (dp[i][r][c] + dp[i-1][r][c+1]) % mod
                }
            }
        }
    }

    return dp[maxMove][startRow][startColumn]
}